Add support for `.nls` extension in NetLogo

danielvartan opened the pull request:
<!--- Briefly describe your changes in the field above. -->

## Description
<!--- If necessary, go into depth of what this pull request is doing. -->

Hi there,

I’d like to propose adding the `.nls` extension to the [NetLogo](https://www.netlogo.org/) language in `linguist`. This is an official NetLogo file type, but it's currently missing from `languages.yml`.

I've added heuristics to differentiate `.nls` files as NetLogo or TeX. Each pattern was tested individually, and all return results associated with the extension. `.nls` files that don't match either the NetLogo or TeX patterns are usually Windows Resource Script (RC) files, which Linguist does not have a dedicated language entry for.

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- Please remove whole sections, not points within the sections, that do not apply -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] **I am adding a new extension to a language.**
  - [x] The new extension is used in hundreds of repositories on GitHub.com
    - Search results for each extension:
      <!-- Replace FOOBAR with the new extension, and KEYWORDS with keywords unique to the language. Repeat for each extension added. -->
      - `.nls` (All | ~6.1k files): https://github.com/search?q=NOT+is%3Afork+path%3A*.nls&type=code&utf8=%E2%9C%93
      - `.nls` (NetLogo patterns | ~2.7k files): <https://github.com/search?q=NOT+is%3Afork+path%3A*.nls+%2F%5E%5Cs*%28to%28-report%29%3F%5Cs%2B%5Cw%5B%5Cw-%5D*%7C%28extension%7Cglobal%7C__include%29s%5Cs*%5C%5B%7C%28turtles%7Cpatches%7Clinks%29-own%5Cs*%5C%5B%7C%28%28un%29%3Fdirected-link-%29%3Fbreed%5Cs*%5C%5B%29%2F&type=code&utf8=%E2%9C%93>
      - `.nls` (TeX patterns | 118 files): <https://github.com/search?q=NOT+is%3Afork+path%3A*.nls+%2F%5E%5Cs*%28%5C%5Cbegin%5C%7B%7C%5C%5Cend%5C%7B%7C%5C%5Cnomgroup%5C%7B%7C%5C%5Citem%29%2F&type=code&utf8=%E2%9C%93>
      - `.nls` (NOT NetLogo | ~3.4K files): <https://github.com/search?q=NOT+is%3Afork+path%3A*.nls+NOT+%2F%5E%5Cs*%28to%28-report%29%3F%5Cs%2B%5Cw%5B%5Cw-%5D*%7C%28extension%7Cglobal%7C__include%29s%5Cs*%5C%5B%7C%28turtles%7Cpatches%7Clinks%29-own%5Cs*%5C%5B%7C%28%28un%29%3Fdirected-link-%29%3Fbreed%5Cs*%5C%5B%29%2F&type=code&utf8=%E2%9C%93%3E>
  - [x] I have included a real-world usage sample for all extensions added in this PR:
    - Sample source(s):
      - `0_init.nls` (NetLogo): <https://github.com/WKSu/compound-events/blob/30a990574b2e2f0fe84b72492ed66ef40844a122/code/0_init.nls>
      - `config-reader.nls` (NetLogo): <https://github.com/swarmfabsim/swarmfabsim.github.io/blob/418686c8eedc9d8d39ef4c6907f3c7690360477f/src/config-reader.nls>
      - `nodes.nls` (NetLogo): <https://github.com/ric-colasanti/TRIFIC/blob/691a8f05e0facf21383df381681d040c9c2e9abd/nodes.nls>
      - `output.nls` (NetLogo): <https://github.com/harrykipper/covid/blob/3593a2e73df8764dad1625a4e3f87fc3384b07dd/output.nls>
      - `police.nls` (NetLogo): <https://github.com/bkalthoff/DistIntSys20/blob/f5eaf221063548a55fee5088c4d41c9e00e7bc2a/police.nls>
      - `roads.nls` (NetLogo): <https://github.com/ric-colasanti/TRIFIC/blob/691a8f05e0facf21383df381681d040c9c2e9abd/Dev/roads.nls>
      - `setup-map.nls` (NetLogo): <https://github.com/sustentarea/logoclim/blob/adcd28753718c143540615cad46b9983524c0d74/nlogox/nls/setup-map.nls>
      - `setup-procedures.nls` (NetLogo): <https://github.com/comses/megadapt/blob/6230bb4710ed4326e1298edf4788d8fb84bcbf20/src/netlogo/setup-procedures.nls#L81>
      - `show-values.nls` (NetLogo): <https://github.com/danielvartan/nlogo-utils/blob/48a61c93da6d8469703b36540abe9b81c6b373b1/nlogo/show-values.nls>
      - `edengths.nls` (TeX): <https://github.com/fneum/ev_chargingcoordination2017/blob/581cd3879af85d269d38f446556a9ea600e87457/docs/latex_dissertation/edengths.nls>
      - `Hauptdatei.nls` (TeX): <https://github.com/PrinceSimple/Seminararbeit_Generative_Modelle/blob/f4cecddea929c4f3e362cc687ad968e58ed754cc/Hauptdatei.nls>
      - `main.nls` (TeX): <https://github.com/ShevonKuan/Engineering-mechanics/blob/c80e4ee494bd9fed191ed590e07e7748d560d666/main.nls>
      - `myThesis.nls` (TeX): <https://github.com/jiec827/njustThesis/blob/bc30a8096b649a79367aa2f0ba29534401463705/myThesis.nls>
      - `tcc-ifbaiano.nls` (TeX): <https://github.com/SavioKSLopes/TCC/blob/ebce4fff70634c0d2503781a9885fbb1f00ccf81/tcc-ifbaiano.nls>
    - Sample license(s):
      - `0_init.nls` (NetLogo): GNU General Public License v3.0 (<https://github.com/WKSu/compound-events/blob/30a990574b2e2f0fe84b72492ed66ef40844a122/LICENSE>)
      - `config-reader.nls` (NetLogo): Creative Commons Attribution 4.0 International (<https://github.com/swarmfabsim/swarmfabsim.github.io/blob/418686c8eedc9d8d39ef4c6907f3c7690360477f/license.txt>)
      - `nodes.nls` (NetLogo): Apache License 2.0 (<https://github.com/ric-colasanti/TRIFIC/blob/691a8f05e0facf21383df381681d040c9c2e9abd/LICENSE>)
      - `output.nls` (NetLogo): GNU General Public License v3 (<https://github.com/harrykipper/covid/blob/3593a2e73df8764dad1625a4e3f87fc3384b07dd/LICENSE>)
      - `police.nls` (NetLogo): MIT License (<https://github.com/bkalthoff/DistIntSys20/blob/f5eaf221063548a55fee5088c4d41c9e00e7bc2a/LICENSE>)
      - `roads.nls` (NetLogo): Apache License 2.0 (<https://github.com/ric-colasanti/TRIFIC/blob/691a8f05e0facf21383df381681d040c9c2e9abd/LICENSE>)
      - `setup-map.nls` (NetLogo): GNU General Public License v3 (<https://github.com/sustentarea/logoclim/blob/adcd28753718c143540615cad46b9983524c0d74/LICENSE.md>)
      - `setup-procedures.nls` (NetLogo): GNU General Public License v3 (<https://github.com/comses/megadapt/blob/6230bb4710ed4326e1298edf4788d8fb84bcbf20/LICENSE>)
      - `show-values.nls` (NetLogo): Creative Commons CC0 1.0 Universal License (<https://github.com/danielvartan/logoutils/blob/48a61c93da6d8469703b36540abe9b81c6b373b1/LICENSE.md>)
      - `edengths.nls` (TeX): GNU General Public License v3.0 (<https://github.com/fneum/ev_chargingcoordination2017/blob/581cd3879af85d269d38f446556a9ea600e87457/LICENSE>)
      - `Hauptdatei.nls` (TeX): MIT License (<https://github.com/PrinceSimple/Seminararbeit_Generative_Modelle/blob/f4cecddea929c4f3e362cc687ad968e58ed754cc/LICENSE>)
      - `main.nls` (TeX): GNU General Public License v3 (<https://github.com/ShevonKuan/Engineering-mechanics/blob/c80e4ee494bd9fed191ed590e07e7748d560d666/LICENSE>)
      - `myThesis.nls` (TeX): GNU General Public License v2 (<https://github.com/jiec827/njustThesis/blob/bc30a8096b649a79367aa2f0ba29534401463705/LICENSE>)
      - `tcc-ifbaiano.nls` (TeX): BSD 3-Clause "New" or "Revised" License (<https://github.com/SavioKSLopes/TCC/blob/ebce4fff70634c0d2503781a9885fbb1f00ccf81/LICENSE>)

danielvartan committed:
Add support for `.nls` extension in NetLogo

lildude reviewed (changes requested):
> Please note that the `.nls` extension is unique—no other language in `languages.yml` uses it.

And that's going to cause a problem as there appear to be more non-NetLogo files than there are NetLogo (add NOT before the keyword in your search) which means they will all be incorrectly classified.

You will need to identity the other main user of this extension and add support at the same time in this PR and use a heuristic to differentiate the two.

danielvartan committed:
Add `.nls` heuristics and map the extension to TeX and INI file types

danielvartan commented:
Hi @lildude,

Thanks for pointing out the issue.

I've added heuristics to distinguish `.nls` files as NetLogo, TeX, or INI, along with two sample files for each type. I also updated the PR description and the checklist sections.

Let me know if any further changes are needed.

lildude committed:
Merge branch 'main' into main

danielvartan committed:
Merge branch 'main' into main

danielvartan committed:
Merge branch 'main' into main

Alhadis reviewed (changes requested):
I question the need for heuristics in the first place, as our classifier should do a decent enough job disambiguating between the filetypes. If not, more (and better-quality) samples are needed.

Regardless, the feedback I've left relates specifically to the accuracy and formatting of the heuristics you've added.

danielvartan committed:
Merge branch 'github-linguist:main' into main

danielvartan committed:
Update `.nls` heuristics and samples

lildude commented on samples/INI/ben.nls:
```diff
@@ -0,0 +1,164 @@
+/*
```
This doesn't look like valid INI. Note everything is wrapped in `{ }` and there are no `=`.

danielvartan replied:
You're right. These `.nls` files are actually locale definition files using Windows Resource Script (RC) syntax. Since Linguist does not have a dedicated RC language, I just removed them. (https://github.com/danielvartan/linguist/commit/abe8651a1957f354f9fc98673a89f6bfdd86e701)

lildude commented on samples/INI/ena.nls:
```diff
@@ -0,0 +1,168 @@
+/*
```
This also doesn't look like INI.

danielvartan replied:
You're right. These `.nls` files are actually locale definition files using Windows Resource Script (RC) syntax. Since Linguist does not have a dedicated RC language, I just removed them. (https://github.com/danielvartan/linguist/commit/abe8651a1957f354f9fc98673a89f6bfdd86e701)

lildude reviewed (changes requested):
See inline comments. Those files need to be correctly identified else they'll introduce false information to the classifier.

samples/INI/j9bcv.nls and samples/NetLogo/parameters.nls are also waaaay too big (if it's suppressed in the diff, it's too big) and we really don't need quite so many samples for each language. 2-3 at most is normally sufficient to train the classifier.

danielvartan committed:
Remove `samples/INI/j9bcv.nls` and `samples/NetLogo/parameters.nls`

danielvartan commented on samples/NetLogo/nodes.nls:
```diff
@@ -0,0 +1,85 @@
+          set thickness 0.1
+          set distance_to_destination link-distance
+          let l-speed item 3 row
+          set speed_limmit 1.0 * 1.0
```
Not applicable. The examples files must match the original files found in GitHub.

danielvartan commented on samples/NetLogo/nodes.nls:
```diff
@@ -0,0 +1,85 @@
+roads-own [
+  r-id ; ID of road
+  distance_to_destination ; Distance to the destination node
+  speed_limmit ; Speed limit of the road
```
Not applicable. The examples files must match the original files found in GitHub.

danielvartan commented on samples/NetLogo/police.nls:
```diff
@@ -0,0 +1,310 @@
+    setxy [ xcor ] of homeAdress [ ycor ] of homeAdress
+    setxy random-xcor random abs 11
+    set next-task [ -> move-police ]
+    ask police [
+      ifelse shift <= number-of-police / 2 [
+        set dayShift? true
+      ] [
+        set dayShift? false
+      ]
+      set shift shift - 1
+    ]
```
Not applicable. The examples files must match the original files found in GitHub.

danielvartan commented on samples/NetLogo/setup-map.nls:
```diff
@@ -0,0 +1,105 @@
+to setup-map
+  assert-gis dataset
+
+  let #envelope gis:envelope-of dataset
+
+  ifelse ((item 0 #envelope < -150) or (item 1 #envelope > 150)) [
+    gis:set-world-envelope-ds #envelope
```
This is wrong. Variables in NetLogo can start with `#`. Example: https://github.com/sustentarea/logoclim/blob/main/nlogox/nls/setup-plot-window.nls.

danielvartan commented on samples/NetLogo/output.nls:
```diff
@@ -0,0 +1,212 @@
+to current-rt
+  ifelse length table:keys infections > 0
+  [
+    let allinfections []
+    foreach table:values infections [dd -> foreach dd [dd1 -> set allinfections lput dd1 allinfections]]
+    if length allinfections > 1 [
+      set rtime mean allinfections
+      let spreaders length all-infections
+      set k0 ifelse-value rtime > 0 [(sum sublist reverse sort all-infections 0 (spreaders * 0.2)) / sum all-infections][0]
```
Not applicable. The example files must match the original files found on GitHub.

danielvartan commented on samples/NetLogo/police.nls:
```diff
@@ -0,0 +1,310 @@
+    set homeAdress one-of houses
+    set list-of-known-gangsters []
+    set list-of-known-hideouts []
+    setxy [ xcor ] of homeAdress [ ycor ] of homeAdress
+    setxy random-xcor random abs 11
```
Not applicable. The example files must match the original files found on GitHub.

danielvartan commented on samples/NetLogo/police.nls:
```diff
@@ -0,0 +1,310 @@
+    set search-tick 0
+    set size 2
+    set follow-tick -1
+    set shift number-of-police
```
Not applicable. The example files must match the original files found on GitHub.

danielvartan commented on samples/TeX/edengths.nls:
```diff
@@ -0,0 +1,314 @@
+
+  \item [{$\omega_{k,t}$}]\begingroup Regulation service provision of EV at household $k$ in time slot $t$\nomeqref {0}
+		\nompageref{xv}
+  \item [{$P^{EV}_{k,t}$}]\begingroup Charging rate of electric vehicel at household $k$ in time slot $t$ (kW)\nomeqref {0}
```
Not applicable. The example files must match the original files found on GitHub.

danielvartan commented on lib/linguist/heuristics.yml:
```diff
@@ -611,6 +611,32 @@ disambiguations:
+- extensions: ['.nls']
+  rules:
+  - language: NetLogo
+    pattern: >-
```
This is wrong. There is nothing wrong with the indentation.

danielvartan committed:
Remove `ben.nls` and `ena.nls`

danielvartan commented on lib/linguist/languages.yml:
```diff
@@ -3250,6 +3250,7 @@ INI:
   - ".dof"
   - ".frm"
   - ".lektorproject"
+  - ".nls"
   - ".prefs"
```
You're right. These `.nls` files are actually locale definition files using Windows Resource Script (RC) syntax. Since Linguist does not have a dedicated RC language, I just removed them. (https://github.com/danielvartan/linguist/commit/abe8651a1957f354f9fc98673a89f6bfdd86e701)

danielvartan committed:
Fix `.nls` INI pattern

danielvartan committed:
Fix conflicts with main branch

danielvartan committed:
Remove conflict fix

danielvartan committed:
Improve `.nls` INI pattern

danielvartan committed:
Improve `.nls` INI pattern (2)

danielvartan committed:
Remove INI relation with `.nls`

danielvartan commented:
@lildude I've resolved all the issues you pointed out.

I decided to remove the INI association with the `.nls` extension. After searching more thoroughly, I realized that most files not matching either the NetLogo or TeX patterns are Windows Resource Script (RC) files, which Linguist does not have a dedicated language entry for. I found fewer than 20 files that are unambiguously INI.

I've also updated the PR description to reflect these changes, along with the current file count.

Please let me know if any further changes are needed.

danielvartan committed:
Merge branch 'main' into main

lildude commented on lib/linguist/heuristics.yml:
```diff
@@ -619,6 +619,24 @@ disambiguations:
+  rules:
+  - language: NetLogo
+    pattern: >-
+      (?x) ^ \s*
```
We require our regexes to be RE2 compliant as they're re-used by go-enry (written in Go) and other tools written in Rust and other languages which don't have PCRE-compatible regex engines. Please make these regexes RE2 compliant.

danielvartan replied:
Thanks for flagging this. The patterns were already RE2 compliant, no lookaheads, backreferences, or other PCRE-only features.

What I did change was split the free-spacing alternation blocks into separate single-line patterns, one per branch, so they're easier to parse consistently across tools without relying on `(?x)` mode.
